src/FsAutoComplete.Core/InlayHints.fs

```diff
@@ -105,42 +104,221 @@ let truncated (s: string) =
   else
     s
 
-let private isNotWellKnownName =
-  let names = Set.ofList [
-    "mapping"
-    "format"
-    "value"
-    "x"
-  ]
-
-  fun (p: FSharpParameter) ->
-  match p.Name with
-  | None -> true
-  | Some n -> not (Set.contains n names)
-
-let inline hasName (p: FSharpParameter) =
-  not (String.IsNullOrEmpty p.DisplayName)
-  && p.DisplayName <> "````"
-
-let inline isMeaningfulName (p: FSharpParameter) =
-  p.DisplayName.Length > 2
-
-let inline doesNotMatchArgumentText (parameterName: string) (userArgumentText: string) =
-  parameterName <> userArgumentText
-  && not (userArgumentText.StartsWith parameterName)
-
-/// </summary>
-/// We filter out parameters that generate lots of noise in hints.
-/// * parameter has a name
-/// * parameter is one of a set of 'known' names that clutter (like printfn formats)
-/// * parameter has length > 2
-/// * parameter does not match (or is an extension of) the user-entered text
-/// </summary>
-let shouldCreateHint (p: FSharpParameter) (matchingArgumentText: string) =
-  hasName p
-  && isNotWellKnownName p
-  && isMeaningfulName p
-  && doesNotMatchArgumentText p.DisplayName matchingArgumentText
+let private createParamHint
+  (range: Range)
+  (paramName: string)
+  =
+  let format p = p + " ="
+  {
+    Text = format (truncated paramName)
+    InsertText = None
+    Pos = range.Start
+    Kind = Parameter
+  }
+let private createTypeHint
+  (range: Range)
+  (ty: FSharpType)
+  (displayContext: FSharpDisplayContext)
+  =
+  let ty = ty.Format displayContext
+  let format ty = ": " + ty
+  {
+    Text = format (truncated ty)
+    InsertText = Some (format ty)
+    Pos = range.End
+    Kind = Type
+  }
+
+module private ShouldCreate =
+  let private isNotWellKnownName =
+    let names = Set.ofList [
+      "value"
+      "x"
+    ]
+
+    fun (p: FSharpParameter) ->
+    match p.Name with
+    | None -> true
+    | Some n -> not (Set.contains n names)
+
+  let private isWellKnownParameterOrFunction 
+    (func: FSharpMemberOrFunctionOrValue)
+    (param: FSharpParameter)
+    =
+    let startsWith (v: string) (fullName: string) =
+      if fullName.StartsWith v then
+        Some ()
+      else
+        None
+    // doesn't differentiate between modules, types, namespaces 
+    // -> is just for documentation in code
+    let (|Module|_|) = startsWith
```

Comment:
since we're going to this these so often, can you make them struct partial active patterns?

Reply:
done

(ahh...no attributes for local functions...)